File: scripts/btlib/default_scale.py
"""Default placement scale shared by CLI inspection and authoring."""

from __future__ import annotations

from typing import Any

LARGE_ASSET_AXIS = 30.0
LARGE_ASSET_TARGET_AXIS = 6.5
MAX_DEFAULT_SCALE = 30.0


def default_drop_scale(asset: dict[str, Any]) -> float:
    manifest_scale = asset.get("default_scale")
    if (
        isinstance(manifest_scale, int | float)
        and not isinstance(manifest_scale, bool)
        and manifest_scale > 0
    ):
        return float(manifest_scale)

    bbox = asset.get("bbox") if isinstance(asset.get("bbox"), list) else [1, 1, 1]
    max_axis = max(abs(float(value or 0)) for value in bbox)
    if max_axis <= 0:
        return 1.0

    if max_axis > LARGE_ASSET_AXIS:
        return min(MAX_DEFAULT_SCALE, LARGE_ASSET_TARGET_AXIS / max_axis)

    return min(MAX_DEFAULT_SCALE, max(1.0, 1.5 / max_axis))

File: editor/default-scale.js
const LARGE_ASSET_AXIS = 30;
const LARGE_ASSET_TARGET_AXIS = 6.5;
const MAX_DEFAULT_SCALE = 30;

export function defaultDropScale(asset) {
  const manifestScale = Number(asset?.default_scale);
  if (Number.isFinite(manifestScale) && manifestScale > 0) {
    return manifestScale;
  }

  const bbox = Array.isArray(asset?.bbox) ? asset.bbox : [1, 1, 1];
  const maxAxis = Math.max(...bbox.map((value) => Math.abs(Number(value) || 0)));
  if (maxAxis <= 0) return 1;

  if (maxAxis > LARGE_ASSET_AXIS) {
    return Math.min(MAX_DEFAULT_SCALE, LARGE_ASSET_TARGET_AXIS / maxAxis);
  }

  return Math.min(MAX_DEFAULT_SCALE, Math.max(1, 1.5 / maxAxis));
}
